### ISSUE TITLE
Intermittent incorrct 403s - authorization policy issues

### ISSUE THREAD

**Issue by nabbott2008** [istio/istio#26607]
**Bug description**

We have this auth policy

```
kubectl -n abc get authorizationpolicy my-service-inbound-auth -o yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  creationTimestamp: "2020-08-10T09:07:54Z"
  generation: 1
  name: my-service-inbound-auth
  namespace: abc
  resourceVersion: "94904626"
  selfLink: /apis/security.istio.io/v1beta1/namespaces/abc/authorizationpolicies/my-service-inbound-auth
  uid: c4da9ba0-3d74-402b-9def-5a8231aa9de4
spec:
  action: ALLOW
  rules:
  - when:
    - key: source.namespace
      values:
      - abc
  - when:
    - key: request.headers[fooauthkey]
      values:
      - ABCD
  selector:
    matchLabels:
      app: my-service
      environment: abc
```

We have another app in the same namespace that is able to access most of the times but others gets a 403.

Requests go:

other-service -> my-service

SUCCESSFUL TRANSACTION:

other-service sending request:
```
message: [2020-08-17T16:45:26.681Z] ode-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "944e44c2-4dad-473c-a99d-164c1b13041d" "my-service:8080" "1.2.3.5:8080" outbound|8080||my-service.abc.svc.cluster.local 1.2.4.5:54054 10.100.198.10:8080 1.2.4.5:40600 - default
```
my-service receiving request:
```
message: [2020-08-17T16:45:26.681Z] "GET /action/1234 HTTP/1.1" 200 - "-" "-" 0 6809 47 47 "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "944e44c2-4dad-473c-a99d-164c1b13041d" "my-service:8080" "127.0.0.1:8080" inbound|8080|http|my-service.abc.svc.cluster.local 127.0.0.1:57178 1.2.3.5:8080 1.2.4.5:54054 outbound_.8080_._.my-service.abc.svc.cluster.local default
```

UNSUCCESSFUL TRANSACTION:

other-service sending request:
not present!

my-service receiving request:
```
message: [2020-08-17T16:58:20.382Z] "GET /action/1234 HTTP/1.1" 403 - "-" "-" 0 19 0 - "-" "node-fetch/1.0 (+https://github.com/bitinn/node-fetch)" "8e7fd848-19e2-4a5b-9c0e-2e89ff24b772" "my-service:8080" "-" - - 1.2.3.5:8080 1.2.4.5:33786 - -
```

It looks like the request didn't go through the istio sidecar in other-service as I don't see any istio-proxy logs for this request.

These logs are missing on the request my-service gets:

```
\"%UPSTREAM_HOST%\" %UPSTREAM_CLUSTER% %UPSTREAM_LOCAL_ADDRESS%
````
and
```
%REQUESTED_SERVER_NAME%
```

[ ] Docs
[ ] Installation
[ ] Networking
[X] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[X] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Every request that comes from other-service should be allowed through auth policy - issue actually seems to be that the other-service pod is sometimes sending requests direct to the pod IP rather than through its istio-proxy

**Steps to reproduce the bug**

We can reproduce the response using wget:

SUCCESSFUL
my-service log
```
[2020-08-18T14:13:57.640Z] "GET /action/1234 HTTP/1.1" 200 - "-" "-" 0 2708 422 422 "-" "Wget" "9803d726-24af-499a-9c62-5785d726b3d8" "my-service:8080" "127.0.0.1:8080" inbound|8080|http|my-service.abc.svc.cluster.local 127.0.0.1:51398 1.2.3.4:8080 1.2.4.5:56106 outbound_.8080_._.my-service.abc.svc.cluster.local default
```
As expected we see this request in the other-service istio-proxy logs

UNSUCCESSFUL - so we just target the my-service pod IP:
my-service log
```
message: [2020-08-18T14:13:03.078Z] "GET /action/1234 HTTP/1.1" 403 - "-" "-" 0 19 0 - "-" "Wget" "fe6acd6d-f61e-4a65-bad3-b9cfcfc16b87" "1.2.3.4:8080" "-" - - 1.2.3.4:8080 1.2.4.5:51654 - -
````
As expected we don't see this other-service istio logs.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

GitVersion:"v1.17.9-eks-4c6976
client version: 1.5.2
egressgateway version: 1.6.2
ingressgateway version: 1.6.2
ingressgateway-public version:
pilot version: 1.6.2
data plane version: 1.6.2 (315 proxies)

**How was Istio installed?**
kubectl

**Environment where bug was observed (cloud vendor, OS, etc)**
EKS

**Comment by rolandkool** [comment]
We're experiencing very similar behavior in our Istio setup: 1.6.5 on GKE 1.16.

In our case we're seeing a (java) application in a pod getting 403s while a curl with the same URL from the same pod succeeds with a 200. Also, debug logging on Envoy shows no outgoing traffic, like it doesn't see it. However, the correct destination pod is targeted so it does seem to do loadbalancing.

We can replicate it using these steps:

Create a file with some http headers:

```
cat a
GET /testap HTTP/1.1
Host: http-echo.test.svc.cluster.local
User-Agent: curl/7.64.0
Accept: */*

```

Now do a netcat of this file:
```
cat a | nc -v http-echo.test.svc.cluster.local 8080
http-echo.test.svc.cluster.local [10.35.4.74] 8080 (http-alt) open
HTTP/1.1 200 OK
x-app-name: http-echo
x-app-version: 0.2.3
date: Tue, 25 Aug 2020 14:20:01 GMT
content-length: 17
content-type: text/plain; charset=utf-8
x-envoy-upstream-service-time: 61
server: envoy

'Hello world'
```

Doing the same with telnet (copy/paste the contents of file `a` into the telnet screen):
```
telnet http-echo.test.svc.cluster.local 8080
Trying 10.35.4.74...
Connected to http-echo.test.svc.cluster.local.
Escape character is '^]'.
GET /testap HTTP/1.1
Host: http-echo.test.svc.cluster.local
User-Agent: curl/7.64.0
Accept: */*

HTTP/1.1 403 Forbidden
content-length: 19
content-type: text/plain
date: Tue, 25 Aug 2020 13:45:22 GMT
server: istio-envoy
x-envoy-decorator-operation: http-echo.test.svc.cluster.local:8080/*

RBAC: access denied
^]
telnet> quit

```

And finally, a simple curl also works:
```
curl http://http-echo.test.svc.cluster.local:8080/testap
```

**Comment by rolandkool** [comment]
Another observation, if you override (via /etc/hosts) the IP the http-echo.test.svc.cluster.local host resolves to to some random IP that doesn't exist, the `nc` and `curl` still work (as expected because envoy does the loadbalancing), but the `telnet` command outputs this:
```
telnet http-echo.skl.svc.cluster.local 8080
Trying 1.3.4.5...
Connected to http-echo.skl.svc.cluster.local.
Escape character is '^]'.
GET /testap HTTP/1.1
Host: http-echo.skl.svc.cluster.local
User-Agent: curl/7.64.0
Accept: */*

Connection closed by foreign host.
```

This seems to indicate it tries to call the IP of the service directly bypassing Envoys routing rules.

**Comment by rolandkool** [comment]
When doing the `nc` like telnet (copy/pasting the contents of file `a` into the screen), it also received a 403.

```
nc http-echo.test.svc.cluster.local 8080
GET /testap HTTP/1.1
Host: http-echo.test.svc.cluster.local
User-Agent: curl/7.64.0
Accept: */*

HTTP/1.1 403 Forbidden
content-length: 19
content-type: text/plain
date: Wed, 26 Aug 2020 07:23:24 GMT
server: istio-envoy
x-envoy-decorator-operation: http-echo.test.svc.cluster.local:8080/*

RBAC: access denied
```

Looking at tcpdump captures, the difference is that when a 403 is returned, the tool is sending a packet for every line that is entered. When a 200 is returned, the request is done in a single packet.

**Comment by rolandkool** [comment]
We've found the cause. The Istio Protocol Detection timeout was too low (100ms).
See https://github.com/istio/istio/issues/20477 . Even though we are 1.6.5, the timeout was still 100ms in stead of the 5s that was mentioned in the attached PR.

We added `values.meshConfig.protocolDetectionTimeout: 5s` and that fixed it for us. No Envoy restarts are required.

**Comment by nabbott2008** [comment]
Increasing
```
protocolDetectionTimeout
```
fixed this for us. 

5s is the default in 1.7.

**Comment by yangminzhu** [comment]
> Increasing
> 
> ```
> protocolDetectionTimeout
> ```
> 
> fixed this for us.
> 
> 5s is the default in 1.7.

Oh, I debugged this and replied somewhere else, sorry I forgot to reply back here. It's because the connection is downgraded to plaintext from mTLS when protocol detection timeout, and this caused it to be rejected by AuthorizationPolicy because the `source.namespace` and `source.principal` is only available under mTLS. 

The authorization policy is doing exactly its job here, bumping the protocol detection timeout should fix this and it's already done in 1.7 as you mentioned.
